engine/gamesys/src/gamesys/scripts/script_buffer.cpp

```diff
@@ -879,6 +880,191 @@ namespace dmGameSystem
+                    values = (void*) luaTableToArray<int16_t>(L, count, valueType);
+                break;
+                case dmBuffer::VALUE_TYPE_INT32:
+                    values = (void*) luaTableToArray<int32_t>(L, count, valueType);
+                break;
+                case dmBuffer::VALUE_TYPE_INT64:
+                    values = (void*) luaTableToArray<int64_t>(L, count, valueType);
+                break;
+                default:
+                    return DM_LUA_ERROR("%s.%s invalid value type supplied: %ld.", SCRIPT_LIB_NAME, SCRIPT_TYPE_NAME_BUFFERMETADATA, valueTypeRaw);
+            }
+
+            dmBuffer::Result r = dmBuffer::SetMetaData(hbuffer, entry_name, (void*) values, count, valueType);
+            free(values);
+
+            if( r != dmBuffer::RESULT_OK )
+            {
+                return DM_LUA_ERROR("%s.%s: cannot set metadata for buffer: %s", SCRIPT_LIB_NAME, SCRIPT_TYPE_NAME_BUFFERMETADATA, dmBuffer::GetResultString(r));
+            }
+        } else {
+            return DM_LUA_ERROR("%s.%s: invalid metadata", SCRIPT_LIB_NAME, SCRIPT_TYPE_NAME_BUFFERMETADATA);
+        }
+
+        return 0;
+    }
+
+    // copies values from array to (empty) lua table on top of the stack
+    template <typename T>
+    static void arrayToLuaTable(lua_State* L, const T* values, uint32_t count, dmBuffer::ValueType valueType) {
+        if (valueType == dmBuffer::VALUE_TYPE_FLOAT32) {
+            for (uint32_t i=0; i<count; i++) {
+                T value = values[i];
+                lua_pushnumber(L, value);
+                lua_rawseti(L, -2, i+1); // lua indices start from 1
+            }
+        } else {
+            for (uint32_t i=0; i<count; i++) {
+                T value = values[i];
+                lua_pushinteger(L, value);
+                lua_rawseti(L, -2, i+1); // lua indices start from 1
+            }
+        }
+    }
+
+    // local metadata, numbertype  = buffer.get_metadata(self.new_buffer, "aabb")
+    static int GetMetadata(lua_State* L)
+    {
+        DM_LUA_STACK_CHECK(L, 2);
+
+        // get buffer
+        dmScript::LuaHBuffer* buffer = dmScript::CheckBuffer(L, 1);
+        dmBuffer::HBuffer hbuffer = UnpackLuaBuffer(buffer);
+
+        // get metadata entry name
+        dmhash_t entry_name = dmScript::CheckHashOrString(L, 2);
+
+        uint32_t count;
+        dmBuffer::ValueType valueType;
+        void* values;
+        dmBuffer::Result r = dmBuffer::GetMetaData(hbuffer, entry_name, &values, &count, &valueType);
+        if ( r == dmBuffer::RESULT_METADATA_MISSING) {
+            lua_pushnil(L);  // nil for metadata entry
+            lua_pushnil(L); // nil for numbertype
+            return 2;
+        }
+        if ( r != dmBuffer::RESULT_OK ) {
+            return DM_LUA_ERROR("%s.%s: error getting metadata for buffer: %s", SCRIPT_LIB_NAME, SCRIPT_TYPE_NAME_BUFFERMETADATA, dmBuffer::GetResultString(r));
+        }
+
+        lua_newtable(L);
+        switch (valueType) {
```

Comment:
Personally, I find these long switch cases a bit tricky to read.
See `CopyStreamInternal` for an example where the switch case has been shortened.
The benefit for me is that the code is easy to view and verify at a glance.
And it minimizes the risk for copy-paste errors even further.

Reply:
yeah, i've seen this pattern in Defold code before.

Actually, the "varying" code was initially moved to a template function for exactly this reason. To make the switch statement simpler. Thus, you end up having this 'case' clause:

```
case dmBuffer::VALUE_TYPE_FLOAT32:
    values = (void*) LuaTableToArray<float>(L, count, valueType);
break;
```
which is not that bad IMO. 

Of course, having a line like:
```
case dmBuffer::VALUE_TYPE_UINT8:      DM_COPY_STREAM(uint8_t); break;
```

is neat, but there is yet another abstraction layer added in between (that of the marco) which adds more complexity. It looks cool, it will save your copy-paste errors but i wouldn't consider it easy to read or understand. Let alone modify the interface.

Having said that, I'll port both switch statements to use this pattern.

Reply:
Well, yes it adds a little bit more complexity, but not much.
I think the conciseness of the way I suggested makes copy paste errors much less likely.
And I think it's easier to read.